dds/DCPS/transport/framework/NetworkAddress.cpp

```diff
@@ -489,84 +530,252 @@ ACE_INET_Addr choose_single_coherent_address(const ACE_INET_Addr& address, bool
+        set6.insert(*it);
       }
     }
 #endif // ACE_HAS_IPV6
-    if (temp.get_type() == AF_INET && !temp.is_multicast()) {
-      VDBG((LM_DEBUG, "(%P|%t) NetworkAddress::choose_single_coherent_address() - "
-                 "Considering Address %C:%d (%C) - ADDING TO IPv4 NORMAL LIST\n",
-                 temp.get_host_addr(), temp.get_port_number(), temp.get_host_name()));
-      set4.insert(temp);
+    if (it->get_type() == AF_INET && !it->is_multicast()) {
+      if (it->is_loopback()) {
+        VDBG((LM_DEBUG, "(%P|%t) choose_single_coherent_address(list) - "
+                   "Considering Address %C:%d (%C) - ADDING TO IPv4 LOOPBACK LIST\n",
+                   it->get_host_addr(), it->get_port_number(), it->get_host_name()));
+        set4_loopback.insert(*it);
+      } else {
+        VDBG((LM_DEBUG, "(%P|%t) choose_single_coherent_address(list) - "
+                   "Considering Address %C:%d (%C) - ADDING TO IPv4 NORMAL LIST\n",
+                   it->get_host_addr(), it->get_port_number(), it->get_host_name()));
+        set4.insert(*it);
+      }
     }
-  } while (copy.next());
+  }
 
 #ifdef ACE_HAS_IPV6
   if (prefer_loopback && !set6_loopback.empty()) {
-    return *set6_loopback.begin();
+    return tie_breaker(set6_loopback, name);
   }
 #endif // ACE_HAS_IPV6
 
   if (prefer_loopback && !set4_loopback.empty()) {
-    return *set4_loopback.begin();
+    return tie_breaker(set4_loopback, name);
   }
 
 #ifdef ACE_HAS_IPV6
-  if (!set6_linklocal.empty()) {
-    return *set6_linklocal.begin();
+  if (prefer_loopback && !set6_linklocal.empty()) {
+    return tie_breaker(set6_linklocal, name);
   }
   if (!set6.empty()) {
-    return *set6.begin();
+    return tie_breaker(set6, name);
   }
   if (!set6_mapped_v4.empty()) {
-    return *set6_mapped_v4.begin();
+    return tie_breaker(set6_mapped_v4, name);
   }
 #endif // ACE_HAS_IPV6
 
   if (!set4.empty()) {
-    return *set4.begin();
+    return tie_breaker(set4, name);
   }
 
 #ifdef ACE_HAS_IPV6
+  if (!set6_linklocal.empty()) {
+    return tie_breaker(set6_linklocal, name);
+  }
   if (!set6_loopback.empty()) {
-    return *set6_loopback.begin();
+    return tie_breaker(set6_loopback, name);
   }
 #endif // ACE_HAS_IPV6
 
   if (!set4_loopback.empty()) {
-    return *set4_loopback.begin();
+    return tie_breaker(set4_loopback, name);
   }
 
-#endif // !(ACE_MAJOR_VERSION < 6 || (ACE_MAJOR_VERSION == 6 && (ACE_MINOR_VERSION < 3 || (ACE_MINOR_VERSION == 3 && ACE_MICRO_VERSION < 1))))
-  ACE_UNUSED_ARG(prefer_loopback);
-  return address;
+  if (!addresses.empty()) {
+    return tie_breaker(addresses, name);
+  }
+
+  return ACE_INET_Addr();
+}
+
+ACE_INET_Addr choose_single_coherent_address(const OPENDDS_STRING& url, bool prefer_loopback)
+{
+  ACE_INET_Addr result;
+
+  if (url.empty()) {
+    return result;
+  }
+
+  // TODO: Fix for literal IPv6 addresses
+  OPENDDS_STRING host_name_str;
```

Comment:
```suggestion
  String host_name_str;
```

You can replace other instances of `OPENDDS_STRING` with `String`.